From 03413338810851c7248d89abf19abccb30356731 Mon Sep 17 00:00:00 2001
From: Ulrich Kramer <u.kramer@sap.com>
Date: Tue, 28 Apr 2020 07:21:48 +0200
Subject: [PATCH] Add environment variables to kpack image CR

Remove select of variables starting with BP

Filter out VCAP_SERVICES from environment variables

Special handling for VCAP_SERVICES
---
 lib/cloud_controller/kpack/stager.rb                | 8 +++++++-
 spec/unit/lib/cloud_controller/kpack/stager_spec.rb | 5 +++--
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/lib/cloud_controller/kpack/stager.rb b/lib/cloud_controller/kpack/stager.rb
index 132b9e52f..b8f7af120 100644
--- a/lib/cloud_controller/kpack/stager.rb
+++ b/lib/cloud_controller/kpack/stager.rb
@@ -86,7 +86,13 @@ module Kpack
 
     def get_environment_variables(staging_details)
       staging_details.environment_variables.
-        except('VCAP_SERVICES').
+        map do |key, value|
+          if key == 'VCAP_SERVICES'
+            value = JSON.generate(value) if value.is_a?(Hash)
+            key = 'CNB_SERVICES'
+          end
+          [key, value]
+        end.
         map { |key, value| { name: key, value: value.to_s } }
     end
 
diff --git a/spec/unit/lib/cloud_controller/kpack/stager_spec.rb b/spec/unit/lib/cloud_controller/kpack/stager_spec.rb
index 146fb97f8..114d4b8e0 100644
--- a/spec/unit/lib/cloud_controller/kpack/stager_spec.rb
+++ b/spec/unit/lib/cloud_controller/kpack/stager_spec.rb
@@ -11,7 +11,7 @@ module Kpack
     )
     }
     let(:package) { VCAP::CloudController::PackageModel.make }
-    let(:environment_variables) { { 'BP_JAVA_VERSION' => '8.*', 'BPL_HEAD_ROOM' => 0, 'VCAP_SERVICES' => { postgres: [] } } }
+    let(:environment_variables) { { 'BP_JAVA_VERSION' => '8.*', 'BPL_HEAD_ROOM' => 0, 'VCAP_SERVICES' => { postgres: [{ name: 'db1' }] } } }
     let(:staging_memory_in_mb) { 1024 }
     let(:staging_disk_in_mb) { 1024 }
     let(:blobstore_url_generator) do
@@ -82,6 +82,7 @@ module Kpack
               env: [
                 { name: 'BP_JAVA_VERSION', value: '8.*' },
                 { name: 'BPL_HEAD_ROOM', value: '0' },
+                { name: 'CNB_SERVICES', value: '{"postgres":[{"name":"db1"}]}' },
               ]
             }
           }
@@ -147,7 +148,7 @@ module Kpack
         end
 
         let(:environment_variables) do
-          { 'VCAP_SERVICES' => 'ignored', 'FOO' => 'BAR' }
+          { 'FOO' => 'BAR' }
         end
 
         before do
-- 
2.25.1

